Infinite Color Engine

awawa-dev opened the pull request:
# Infinite Color Engine & New Smoothing Algorithms

I’m pleased to present a feature I’ve been looking forward to for a long time and have worked hard to bring to life:  
the **Infinite Color Engine**, along with completely new smoothing algorithms.  
This has been on the roadmap for quite a while, but other priorities had to come first.

---

## What is the Infinite Color Engine?

In short, it marks a **clean break from the old 24-bit color space**, moving to **floating-point precision** while still maintaining performance.  

You might ask: *Why bother, if most sources don’t provide more than 24-bit color, except perhaps the P010 format?*  

Well, here’s the thing:  
such sources have always been available—even with something as basic as a cheap MS2109 grabber or flatbuffers external source.  
In our application, when averaging colors for each LED, we’ve always had access to a high-precision theoretical color space.  
Until now, that precision was being **brutally truncated** to 24-bit.

Another hidden “source” of precision has long been **smoothing**, which interpolates intermediate colors.  
This entire subsystem has now been rewritten to handle **floating-point numbers** at both input and output.  

---

## What Does This Mean in Practice?

-  **The obvious: using the possibilities of lamps that offer support for an extended color palette**  
- Reduced banding, noise, and flicker  
- More stable color transitions across frames  
- Smoother post-processing corrections  

Previously, when the final averaged color happened to fall on the boundary between two integers, even tiny changes in the source could trigger sudden jumps between frames.  
By switching to floating-point, this issue has been **drastically minimized**.

---

## Hardware Support

Extended color support is now available in drivers for:

- All Philips Hue lamps using the **Entertainment API (v1 & v2)**  
- **HD108**  

This covers a wide range of popular devices.  
On my own Philips Hue lamps, the first noticeable improvement was **much smoother transitions**—  
where before there were obvious “steps” in color changes, now the shift feels more continuous.

---

## Why “Infinite”?

The name isn’t just marketing.  
The Infinite Color Engine works with a **color palette of 1.24 × 10²⁷ possible values**.  
For comparison:  
- Old 24-bit color: **1.67 × 10⁷ values**  
- New floating-point engine: **1.24 × 10²⁷ values**  

The difference is, quite literally, astronomical. 🌌

---

## New smoothing algorithms

In addition to the transition to floating-point numbers, the focus was on adding parameters that control the abruptness of color changes (brightness limiter or averaging with the previous frame) and adding the ability to work in the YUV space, not just RGB.

| Title | Description |
|---------|-------------|
| Linear Interpolator | The Linear Interpolator is an adaptation of a legacy algorithm from previous HyperHDR versions, rewritten to use floating-point arithmetic for higher precision. It works by linearly interpolating the current color towards a target color, ensuring the transition completes over a defined duration regardless of the frame rate. Its key characteristic is the ability to smoothly retarget mid-animation, initiating a new, full-duration transition from its current state toward the new goal. |
| RGB Infinite Interpolator | This algorithm smoothly animates RGB colors over a set duration, offering two distinct modes for the transition. The first is a direct linear interpolation for a straight path between colors, while the second is a smoothed mode where the current color gracefully 'chases'' the target to create an ease-in/ease-out effect. A key feature is its ability to intelligently rescale an animation's duration when interrupted, ensuring a perceptually constant speed of change. |
| YUV Infinite Interpolator | This algorithm smoothly interpolates colors by operating in the YUV color space for more perceptually uniform transitions. Its main feature is limiting the rate of luminance change in each step, preventing sudden, jarring flashes of brightness. This ensures a visually pleasing effect, even if it extends the animation beyond its initially set duration to maintain that smoothness. |
| Hybrid Physics Infinite Interpolator | This algorithm smoothly transitions between colors using a hybrid physical model. A linear 'pacer' defines the direct path and timing to the target color, while the actual output color follows this pacer like an object attached to a damped spring. This two-part approach creates fluid, natural-looking animations with customizable inertia and overshoot, all while operating in the perceptually-uniform YUV color space. |
| Exponential Infinite Interpolator | Classic exponential implementation of smoothing updates LED colors toward a target, reacting quickly to large differences and slowing as they approach the target, producing smooth, natural ambient lighting transitions. |

---

## Processing Pipeline (Before → Now)

⬇️ **1. Video Frame Area Averaging for LED**
- **Before:** color (float) truncated to 24-bit at output  
- **Now:** output remains as float  

⬇️ **2. Post-Processing**
- **Before:** received, processed in pipeline, and output 24-bit color. some functions, such as changing luminance or saturation, could use float, but they still received and output 24-bit color  
- **Now:** both processing and input/output work on float colors  

⬇️ **3. Smoothing**
- **Before:** input/output in 24-bit color, internal processing on float  
- **Now:** input, processing, and output all on float colors  

⬇️ **4. LED Drivers**
- **Before:** always received and rendered 24-bit color  
- **Now:** if the device supports it (Philips Hue, HD108) → rendering at higher precision;  
  otherwise → fallback to 24-bit rendering  

---

## The post-processing Pipeline (Before → Now)

#### Old approach (24-bit pipeline)  
- Operated directly on **8-bit nonlinear sRGB values**.  
- Every transformation (brightness, per-channel gamma, calibration, temperature, backlight) worked on already quantized integers → **rounding/precision loss** at every step.  
- **Per-channel gamma correction** was not equivalent to standard gamma adjustment, and could distort hues 😬.  
- Input being **nonlinear** made operations like calibration and saturation adjustment less accurate (they are meant to be done in linear space).  

#### New approach (float3 pipeline)  
- Operates first in **linear sRGB space with float precision** → avoids cumulative rounding errors.  
- **Color temperature applied first**: tinting works best before calibration so that calibration LUT/matrix can correct consistently.  
- **Calibration in colorspace applied in linear domain** → mathematically correct, ensures LUT/matrix is used as designed.  
- **Scale color output** (user parameter)  
- **Convert back to nonlinear sRGB only once**: avoids repeated conversions and precision loss.  
- **Global user gamma correction applied consistently on all channels** → perceptually uniform adjustment ✅.  
- **Brightness and saturation after calibration**: adjusting perceptual properties after ensuring primaries are correct.  
- **Minimal backlight last**: guarantees final safety floor regardless of prior adjustments.  
- **Limit power output**:  scale down only if the power limit (user parameter) is exceeded

👉 Overall: the new pipeline is much more **accurate, perceptually consistent, and scientifically correct**. The old one was simpler but mathematically flawed.  

| Old approach (24-bit) | New approach (float3) |
|----------------------------------------|-------------------------------------|
| **Input**: sRGB ***nonlinear*** 😬 (24-bit).<br>All further steps already lose accuracy due to rounding 😬 | **Input**: sRGB *linear* (float3).<br>Correct starting point for calibration & math ✅ |
| **Steps**:<br>1. Apply Brightness & Saturation<br>2. Apply User Gamma **per channel** 😬<br>3. Calibrate in Colorspace (on nonlinear data 😬)<br>4. Apply Color Temperature<br>5. Apply Minimal Backlight | **Steps**:<br>1. Apply Color Temperature (before calibration ✅)<br>2. Calibrate in Colorspace (linear domain ✅)<br>3. Scale color output (user parameter ✅)<br>4. Convert to nonlinear sRGB (only once ✅)<br>5. Apply User Gamma (global, consistent ✅)<br>6. Apply Brightness & Saturation (perceptual ops ✅)<br>7. Apply Minimal Backlight (final floor ✅) <br>8. Limit power output (scale down only if the power limit is exceeded ✅)|
| **Output**: sRGB *nonlinear* (24-bit). | **Output**: sRGB *nonlinear* (float3 → quantized later). |


---

## And One More Thing: Modernizing the Codebase

Since we recently moved to **Qt 6.8.3** or higher  
(remember, Qt 5.15—released over a decade ago—is now end-of-life),  
the next natural step in this PR is adopting **C++20**.  

This isn’t just about upgrading a version number:  
it brings us **better language features** that improve readability, safety, and overall code quality.  

---

## Known issues

- Since the post-processing flow has been verified and rewritten in accordance with the correct processing order, it may happen that if you had a specific color calibration set, you will have to check it and possibly correct it to suit the current pipeline.

- The configuration of post-processing per LED range was removed because it was complicating the code significantly and is unlikely to return, or at least I don't see any reason to migrate it

- <s>If you used or compiled HyperHDR code under Qt5 before, it may no longer work: Qt5.15 has very poor support for C++20.</s> so far we are good with Qt5.15

satgit62 commented:
> Let's give Qt5.15 a chance. If the compilation process breaks, we'll assess whether it's worth creating a workaround for Qt5.15, or whether, for example, the cross-compilation package should be updated to Qt6.8. I can help you cross-compile this Qt6.8 package, although it will probably lead to a split with webos homebrew: another matter is that the version there has long been obsolete, and your repo offers newer version and many more out-of-the-box features, e.g., ready-made LUT tables.

Hi, thank you. I would appreciate your help cross-compiling this Qt6.8 package if it doesn't work with Qt5.15. Yes, I know that webOS homebrew is not kept up to date, which is why I always follow it and always compile the latest packages. I then distributed the self-compiled versions to the webOS community.

awawa-dev commented:
Hi @satgit62 After few small changes I've managed to compile HyperHDR on Windows using 5.15. I've updated the PR so could you test it again? Maybe my concerns about mixing Qt5.15 with C++20 were exaggerated, although often gcc on Linux is the real functional test ;-)

satgit62 commented:
@awawa-dev , thanks to the change “Infinite Color Engine: Compatibility with Qt5.15,” it could be compiled for webOS without any problems. I will test it and give my feedback tomorrow. Thank you.

NatyaSadella commented:
Nice work Awawa, I've tried it and the new smoothing is immediately noticable. 

Do we need to re-calibrate our LUT's for this new color engine?

satgit62 commented:
The question about temperature settings is valid and would also interest me, but I think everyone should select the temperature of the LEDs accordingly. In the HyperSerial or HyperSerialPico driver, the RGB “white channel aspect” can be balanced. With WLED, this can be balanced using the “Temperature Custom Adjust” function.
However, @awawa-dev  should confirm this, as it is only my interpretation.
Please also note the issues known to awawa:
“Since the post-processing workflow has been reviewed and rewritten according to the correct processing order, you may need to check a specific color calibration you had set and possibly adjust it to the current pipeline.”

awawa-dev commented:
Hi
I've updated the description after getting an independent, outside "review" 😉 of our new post-processing path. I'm open to adding new settings, as long as they don't break the current calibration path.

For brightness, you can adjust it in two places:

-    The Brightness(Luminance) & Saturation step, which uses HSL (Hue, Saturation, Luminance). Here, 'L' is for luminance/brightness.

-    The temperature setting. If red equals green and blue, you are essentially adjusting the scale, not the temperature itself.

I hope this explains how color processing worked previously and how it works now.

---

## The post-processing Pipeline (Before → Now)

#### Old approach (24-bit pipeline)  
- Operated directly on **8-bit nonlinear sRGB values**.  
- Every transformation (brightness, per-channel gamma, calibration, temperature, backlight) worked on already quantized integers → **rounding/precision loss** at every step.  
- **Per-channel gamma correction** was not equivalent to standard gamma adjustment, and could distort hues 😬.  
- Input being **nonlinear** made operations like calibration and saturation adjustment less accurate (they are meant to be done in linear space).  

#### New approach (float3 pipeline)  
- Operates first in **linear sRGB space with float precision** → avoids cumulative rounding errors.  
- **Color temperature applied first**: tinting works best before calibration so that calibration LUT/matrix can correct consistently.  
- **Calibration in colorspace applied in linear domain** → mathematically correct, ensures LUT/matrix is used as designed.  
- **Brightness and saturation after calibration**: adjusting perceptual properties after ensuring primaries are correct.  
- **Convert back to nonlinear sRGB only once**: avoids repeated conversions and precision loss.  
- **Global gamma correction applied consistently on all channels** → perceptually uniform adjustment ✅.  
- **Minimal backlight last**: guarantees final safety floor regardless of prior adjustments.  

👉 Overall: the new pipeline is much more **accurate, perceptually consistent, and scientifically correct**. The old one was simpler but mathematically flawed.  

| Old approach (24-bit nonlinear sRGB) | New approach (float3, linear sRGB) |
|----------------------------------------|-------------------------------------|
| **Input**: sRGB *nonlinear* (24-bit).<br>All further steps already lose accuracy due to rounding 😬 | **Input**: sRGB *linear* (float3).<br>Correct starting point for calibration & math ✅ |
| **Steps**:<br>1. Apply Brightness & Saturation<br>2. Apply User Gamma **per channel** 😬<br>3. Calibrate in Colorspace (on nonlinear data 😬)<br>4. Apply Color Temperature<br>5. Apply Minimal Backlight | **Steps**:<br>1. Apply Color Temperature (before calibration ✅)<br>2. Calibrate in Colorspace (linear domain ✅)<br>3. Apply Brightness & Saturation (perceptual ops after calibration ✅)<br>4. Convert to nonlinear sRGB (only once ✅)<br>5. Apply User Gamma (global, consistent ✅)<br>6. Apply Minimal Backlight (final floor ✅) |
| **Output**: sRGB *nonlinear* (24-bit). | **Output**: sRGB *nonlinear* (float → quantized later). |

awawa-dev commented:
Sorry, I missed that question earlier: “Smoothing Factor” is only for the RGB Infinite interpolator (one of the new Smoothing algos).

satgit62 commented:
> Hi I've updated the description after getting an independent, outside "review" 😉 of our new post-processing path. I'm open to adding new settings, as long as they don't break the current calibration path.
> 
> For brightness, you can adjust it in two places:
> 
>     * The Brightness(Luminance) & Saturation step, which uses HSL (Hue, Saturation, Luminance). Here, 'L' is for luminance/brightness.
> 
>     * The temperature setting. If red equals green and blue, you are essentially adjusting the scale, not the temperature itself.
> 
> 
> I hope this explains how color processing worked previously and how it works now.
> ## The post-processing Pipeline (Before → Now)
> #### Old approach (24-bit pipeline)
> 
>     * Operated directly on **8-bit nonlinear sRGB values**.
> 
>     * Every transformation (brightness, per-channel gamma, calibration, temperature, backlight) worked on already quantized integers → **rounding/precision loss** at every step.
> 
>     * **Per-channel gamma correction** was not equivalent to standard gamma adjustment, and could distort hues 😬.
> 
>     * Input being **nonlinear** made operations like calibration and saturation adjustment less accurate (they are meant to be done in linear space).
> 
> 
> #### New approach (float3 pipeline)
> 
>     * Operates first in **linear sRGB space with float precision** → avoids cumulative rounding errors.
> 
>     * **Color temperature applied first**: tinting works best before calibration so that calibration LUT/matrix can correct consistently.
> 
>     * **Calibration in colorspace applied in linear domain** → mathematically correct, ensures LUT/matrix is used as designed.
> 
>     * **Brightness and saturation after calibration**: adjusting perceptual properties after ensuring primaries are correct.
> 
>     * **Convert back to nonlinear sRGB only once**: avoids repeated conversions and precision loss.
> 
>     * **Global gamma correction applied consistently on all channels** → perceptually uniform adjustment ✅.
> 
>     * **Minimal backlight last**: guarantees final safety floor regardless of prior adjustments.
> 
> 
> 👉 Overall: the new pipeline is much more **accurate, perceptually consistent, and scientifically correct**. The old one was simpler but mathematically flawed.
> Old approach (24-bit nonlinear sRGB) 	New approach (float3, linear sRGB)
> **Input**: sRGB _nonlinear_ (24-bit).
> All further steps already lose accuracy due to rounding 😬 	**Input**: sRGB _linear_ (float3).
> Correct starting point for calibration & math ✅
> **Steps**:
> 1. Apply Brightness & Saturation
> 2. Apply User Gamma **per channel** 😬
> 3. Calibrate in Colorspace (on nonlinear data 😬)
> 4. Apply Color Temperature
> 5. Apply Minimal Backlight 	**Steps**:
> 1. Apply Color Temperature (before calibration ✅)
> 2. Calibrate in Colorspace (linear domain ✅)
> 3. Apply Brightness & Saturation (perceptual ops after calibration ✅)
> 4. Convert to nonlinear sRGB (only once ✅)
> 5. Apply User Gamma (global, consistent ✅)
> 6. Apply Minimal Backlight (final floor ✅)
> **Output**: sRGB _nonlinear_ (24-bit). 	**Output**: sRGB _nonlinear_ (float → quantized later).

Your explanation was a veritable seminar on what functions and algorithms in HyperHDR can achieve, both old and new. I must admit that even at my advanced age, I am happy to learn and internalize new things. Thank you for that. 🙏🫡 👍

awawa-dev commented:
@NatyaSadella 
> Do we need to re-calibrate our LUT's for this new color engine?

No, you don't.

> Edit: I see new settings in the image processing tab. About the color temperature settings: Is this to "correct" a SK6812 neutral-white version to your recommended cold-white version? Or is this about adjusting your LED's to match the white point of the TV picture? Almost everywhere a D65 white point is recommended, is this setting supposed to match that?

I think it should be disabled by default. In our case, it's a completely subjective option and depends on how you want the white shifted in processing, which of course affects the output. Or/and you can use the HyperSerial/HyperSPI option to calibrate the output colors, which works differently: simultaneous calibration of white color and temperature. By the way, in the HyperSerial/HyperSPI , the 'cold' setting has equal correction factors (something like setting the same value for all channels here for temperature) because all SK6812 LEDs on the market that I met already have a red shift, and in fact, the SK6812 cold is 'neutral' here, and there's no such thing as a "real-cold white" SK6812.

awawa-dev commented:
New features:
 - The Hue lamp HyperHDR driver  has been significantly rewritten. Protection against inaccuracies when calculating very dark colors has been added to the algorithms. "Candy gamma" is now permanently enabled, as it should be. A potential bug in the wizard has been potentially fixed: #1268  

- Processing (the entire color calibration pipeline) was previously disabled for colors and effects; it is now disabled only for colors.

- Processing: Added the **"Scale color output"** option (multiplies colors by a user-specified factor).

- Processing: Added the **"Limit power output"** option. This works a bit like scaling, but dynamically when a given frame exceeds a specified limit. Please do not confuse this with a power limiter, where you can set a specific value in watts. We use an abstract 0-1 scale, and it only regulates the maximum brightness of scenes.

the doc has been updated

awawa-dev commented:
Many articles state similar to "HSV is a perceptual color space, so you should feed it perceptual RGB". So be it, let's move applying user luminance and saturation setting after gamma (sRGB and user) correction.

BTW, to be 100% compliant with the Hue API requirements, our user gamma should be 1.0: so it should be disabled, use only sRGB colorspace which gamma gives us non-linear space, but later in Hue driver we restore it back to linear before sending because we need convert it to xyBrightness coords (theoretically we can send RGB in newer v2 API and let Hue bridge to do the conversion, but it doesnt work in v1). User gamma will change a bit this characteristic. But of course, you can use it however you want; first and foremost, it has to suit you.

satgit62 commented:
Hello @awawa-dev,
Is this the “Holy Grail” of color management for HUE lamps?🤩
I have been following this with great interest and am delighted with every improvement you add. I tested it right away and it works.
However, as I don't own any Hue LED lamps or Gradient Stripes with extended color space, I am unfortunately unable to form my own opinion.
My Philips Bloom and Ledvance/Osram bulbs work perfectly, but to be honest, I don't know what to look for to notice the differences. I've read everything carefully, but somehow I'm still none the wiser.🤔
I would appreciate a simple explanation that I can pass on to others. 😊

awawa-dev commented:
I have two White and Color Ambiance Play lamps under the TV, just below the home cinema speakers. They're even more noticeable than the LED strip behind the TV, so I have to take care of them ;-) The Hue API is perfectly described with documentation available (you just need to register with them), so the expected color space follows directly from it. 
BTW I see room for further improvement or testing: let's move smoothing behind post-processing? Or consider unifying the sRGB (2.2) and user gamma, controlling it with a single parameter. But that's tomorrow.

awawa-dev commented:
⚠️ BREAKING CHANGE ⚠️ 

Smoothing works best on colors which the users sees (on LEDs, perceptual , nonlinear) so let's put smoothing after post-processing. Fortunately changing the order was an easy fix.

awawa-dev commented:
⚠️ BREAKING CHANGE ⚠️ 

Because after moving brightness to its proper place, we suddenly have no way to adjust the scaling of values ​​in linear space after calibration and before conversion to linear. I've moved the "Color scale" there just before returning to the non-linear space: nmo side effects just a different behaviuor. I hope this is the end of the processing optimization and we can start testing smoothing. the doc is updated.

⚠️ SMOOTHING ⚠️ 

Testing smoothing is quite a challenge. What works on paper, assuming we set a goal and a set time and nothing will interfere with it, does not work in the environment where the grabber is constantly setting a new target at a high frequency and the LED clock is constantly querying for current colors also using high frequence.

But we will deal with it with time. For now, I've ported the coefficients from the old algorithm, so linear smoothing should work like HyperHDR Advanced Smoothing, not the old Hyperion algorithm without them.

So that everyone can see how the algorithm works in practice with the same input data, a test mode has been added as a checkbox in the smoothing settings. When enabled, smoothing is fed a circular pattern with the following color sequence:

```
			{0,    0.0f, 0.0f, 0.0f},  // start from black color
			{2000, 1.0f, 0.0f, 0.0f},  // short (100ms) red color blink at 2.0sec
			{2100, 0.0f, 0.0f, 0.0f},  // return to black color
			{2300, 0.0f, 1.0f, 0.0f},  // short (100ms) green color blink at 2.3sec
			{2400, 0.0f, 0.0f, 0.0f},  // return to black color
			{2600, 0.0f, 0.0f, 1.0f},  // short (100ms) blue color blink at 2.6sec
			{2700, 0.0f, 0.0f, 0.0f},  // return to black color
			{3000, 0.0f, 0.0f, 1.0f},  // start moving to blue color at 3sec
			{3200, 0.0f, 1.0f, 0.0f},  // then turn to green color at 3.2sec
			{3400, 1.0f, 0.0f, 0.0f},  // then immediately turn to red color at 3.45sec and wait
			{4000, 0.0f, 0.0f, 1.0f},  // start moving to blue color at 4sec
			{4100, 1.0f, 0.0f, 0.0f},  // then turn to red color at 4.1sec
			{4200, 0.0f, 1.0f, 0.0f},  // the turn to green at 4.2sec and wait 
			{6000, 0.0f, 0.0f, 0.0f},  // finally return to black color at 6sec
			{8000, 0.0f, 0.0f, 0.0f}   // the end of the cycle at 6sec
```	
	
	
then you will see on the console output (on WIndows you must run hyperhdr.exe -c from the command line):

<img width="1348" height="766" alt="obraz" src="https://github.com/user-attachments/assets/3937f53f-5ad8-4cc2-aa77-034e41392d3a" />

here you have a moment from the sequence when smoothing is starting going from the green color to the black color:
```
			{4200, 0.0f, 1.0f, 0.0f},  // the turn to green at 4.2sec and wait 
			{6000, 0.0f, 0.0f, 0.0f},  // finally return to black color at 6sec
```

However, the result will be slightly different for everyone because, as I mentioned, the smoothing algorithm depends on the frequency of color delivery (video grabber capturing rate) and the LED refresh rate.

I hope this helps explain how these algorithms behave in practice, and how difficult it is to maintain a given refresh rate (in fact, it's impossible unless we compromise on smoothing quality). Perhaps you can even help me improve these new smoothing algorithms with your ideas 😃 And today anyone can be a Vibe Coder, so there's no harm in trying 🚀

satgit62 commented:
Hello @awawa-dev,

I don't know what changes were made in the Philips driver version, but I often have problems with the handshake starting and timeouts under LG webOS. I have to reset Philips Hue to “file” and reconfigure it every time. Other users who use this branch have also noticed this.

When I recompile a newly added feature for webOS – as I did today with the Infinite Color Engine and the new Exponential Smoothing feature – the daemon no longer starts with the old hyperhdr.db.
It runs again when I delete hyperhdr.db and reconfigure everything.
If I only configure Adalight, there are no startup problems and it also runs with the old hyperhdr.db. So the problem must be related to Philips Hue.

I also noticed that after the new commit (Add new Exponential smoothing), the luminance setting is missing in the UI.
However, it still works via LG Input Hook or JSON Link in a normal browser (I configured it a few days ago when it was still available in the UI).

satgit62 commented:
Hello @awawa-dev,
After restarting, Philips Hue had problems again. I started it from the command line as suggested. Here are the results:
I hope you can find something that will help us.

[root@LGwebOSTV~# LD_LIBRARY_PATH=me.txt](https://github.com/user-attachments/files/22234325/root%40LGwebOSTV.LD_LIBRARY_PATH.me.txt)

satgit62 commented:
I think I've found something that might be causing the problem:
After turning off my Home Assistant and restarting the TV, the Hue lamps work as intended again.
I'll turn Home Assistant back on and see if it causes this error.
And yes, my lamps are also configured there to.
EDIT:
Yes, it's related to Home Assistant. When I turn it back on, I encounter problems again when configuring “philipshue” as an LED controller. I have no idea why these problems occur.

HAWK77766 commented:
YUV interpolator is the smoothest. It outperforms the others algorithms in the near black zone.
I have great results at these settings: 25ms, 143hz, Brightness change limiter 0,001

Had one not reproducible crash trying diffrent settings with the YUV interpolator.

awawa-dev commented:
@HAWK77766
> Had one not reproducible crash trying different settings with the YUV interpolator.

Thank you, I think I managed to pin point the issue and fixed it in the latest commit.

@satgit62 
I wonder if that has something to do with QT6.8 OR OpenSSL3.x

awawa-dev commented:
Added Infinity Hybrid Interpolator...It seems to work, or at least it's "dancing", when the direction changes suddenly: for a short moment, it continues moving relative to the old target. For example: it was moving towards green color target, then suddenly target changes to black color, and the green color still continues to increase for a moment despite that.

All that remains is to choose some sensible default parameters; for now, they are what they are. Need more testing.

Since the question was asked earlier how these parameters (stiffness,damping) affect hybrid interpolation, here we can practice it.
[link](https://www.myphysicslab.com/springs/single-spring-en.html?DAMPING%3D13%3BMASS%3D1%3BSPRING_STIFFNESS%3D150%3BDIFF_EQ_SOLVER%3DEULERS_METHOD%3BTIME_STEP%3D0.02%3BFIRING%3Dtrue%3BPOSITION%3D-1.9392177078577043e-12%3BVELOCITY%3D-6.191198557663425e-12%3BTIME%3D263.0500000000698%3BSIN_TIME%3D-0.7471627320125664%3B)

satgit62 commented:
> @satgit62 I wonder if that has something to do with QT6.8 OR OpenSSL3.x

Hi awawa-dev,
I wanted to let you know that the option “1 Test OpenSSL 1.1, stay on QT6.8” was successful.👍 

It didn't work with the buildroot-nc4 (https://github.com/openlgtv/buildroot-nc4/releases/download/webos-c592d84/arm-webos-linux-gnueabi_sdk-buildroot.tar.gz) because some things hadn't been updated in a long time. But I compiled my own buildroot-nc4 some time ago, which I used to compile all your branches, and it worked this time too.

With OpenSSL 1.1, it worked right away without me having to change any settings. To perform a meaningful test, I even deleted the hyperhdr.db completely and used my old Hyperhdr.db. The daemon was restarted – and lo and behold, the lights came on as they should!!!!😃
So the problem was with OpenSSL3; with the old hyperhdr.db, the daemon didn't start at all.

<img width="1623" height="770" alt="OpenSSL 1 1 1q" src="https://github.com/user-attachments/assets/b85b1536-087f-4141-b0cb-a007356d3d5b" />

satgit62 commented:
In the meantime, I've added a smoothing factor to the YUV interpolator. And Hybrid smoothing should be working too. That's all for now. Have fun testing :-)

That's great work, exceeding all my expectations!👍 Thank you.

satgit62 commented:
@awawa-dev ,
As far as I can tell, you have pushed a new commit that affects the Infinite Color Engine (stronger YUV target achievement detection limit (3)). Thank you for that.
